Documentation/Contributors/CodingGuide/README.md

```diff
@@ -0,0 +1,662 @@
+    ], function(
+        freezeObject) {
+    "use strict";
+
+    var ModelAnimationState = {
+        STOPPED : 0,
+        ANIMATING : 1
+    };
+
+    return freezeObject(ModelAnimationState);
+});
+```
+* Use descriptive comments for non-obvious code, e.g.,
+```javascript
+byteOffset += sizeOfUint32; // Add 4 to byteOffset
+```
+is better written as
+```javascript
+byteOffset += sizeOfUint32; // Skip length field
+```
+* `TODO` comments need to be removed or addressed before the code is merged into master.  Used sparingly, `PERFORMANCE_IDEA`, can be handy later when profiling.
+
+## Functions
+
+* :art: Functions should be **cohesive**; they should only do one task.
+* Statements in a function should be at a similar level of abstraction.  If a code block is much lower level than the rest of the statements, it is a good candidate to move to a helper function, e.g., 
+```javascript
+Cesium3DTileset.prototype.update = function(frameState) {
+    var tiles = this._processingQueue;
+    var length = tiles.length;
+
+    for (var i = length - 1; i >= 0; --i) {
+        tiles[i].process(this, frameState);
+    }
+
+    selectTiles(this, frameState);
+    updateTiles(this, frameState);
+};
+```
+is better written as
+```javascript
+Cesium3DTileset.prototype.update = function(frameState) {
+    processTiles(this, frameState);
+    selectTiles(this, frameState);
+    updateTiles(this, frameState);
+};
+
+function processTiles(tiles3D, frameState) {
+    var tiles = tiles3D._processingQueue;
+    var length = tiles.length;
+
+    for (var i = length - 1; i >= 0; --i) {
+        tiles[i].process(tiles3D, frameState);
+    }
+}
+```
+* :speedboat: Smaller functions are more likely to be optimized by V8.  Consider this for code that is likely to be a hot spot.
+* :speedboat: Avoid multiple `return` statements in small hot functions since V8 will not inline them .
+
+### `options` Parameters
+
+:art: Many Cesium functions take an `options` parameter to support optional parameters, self-documenting code, and forward compatibility.  For example, consider:
+```javascript
+var sphere = new SphereGeometry(10.0, 32, 16, VertexFormat.POSITION_ONLY);
+```
+It is not clear what the numeric values represent and the caller needs to know the order of parameters.  If this took an `options` parameter, it would look like:
+```javascript
+var sphere = new SphereGeometry({
+    radius : 10.0,
+    stackPartitions : 32,
+    slicePartitions : 16,
+    vertexFormat : VertexFormat.POSITION_ONLY
+});
+```
+* :speedboat: Using `{ /* ... */ }` creates an object literal, which is a memory allocation.  Avoid designing functions that use an `options` parameter if the function is likely to be a hot spot; otherwise, callers will have to use a scratch variable (see [below](#result-parameters-and-scratch-variables)) for performance.  Constructor functions are good candidates for `options` parameters since Cesium avoids constructing objects in hot spots.
```

Comment:
Maybe provide an example of a function that does not use an options object (any of our Core types like Cartesian3 and Quaternion)